cola/widgets/submodules.py

```diff
@@ -0,0 +1,131 @@
+
+
+class SubmodulesWidget(QtWidgets.QWidget):
+
+    def __init__(self, context, parent):
+        QtWidgets.QWidget.__init__(self, parent)
+
+        self.setToolTip(N_('Submodules'))
+        # main
+        self.tree = SubmodulesTreeWidget(context, parent=self)
+        self.setFocusProxy(self.tree)
+        self.main_layout = qtutils.vbox(
+            defs.no_margin, defs.spacing, self.tree)
+        self.setLayout(self.main_layout)
+        # button
+        # TODO better icons
+        self.refresh_button = qtutils.create_action_button(
+                tooltip=N_('Refresh'), icon=icons.sync())
+        qtutils.connect_button(self.refresh_button, self.tree.refresh_command)
+        self.back_to_parent_button = qtutils.create_action_button(
+                tooltip=N_('GoToParent'), icon=icons.push())
+        qtutils.connect_button(self.back_to_parent_button,
+                               self.tree.back_to_parent_command)
+        self.button_layout = qtutils.hbox(defs.no_margin, defs.spacing,
+                                          self.back_to_parent_button,
+                                          self.refresh_button)
+        self.corner_widget = QtWidgets.QWidget(self)
+        self.corner_widget.setLayout(self.button_layout)
+        # titlebar
+        titlebar = parent.titleBarWidget()
+        titlebar.add_corner_widget(self.corner_widget)
+
+
+class SubmodulesTreeWidget(standard.TreeWidget):
+    updated = Signal()
+
+    def __init__(self, context, parent=None):
+        standard.TreeWidget.__init__(self, parent=parent)
+
+        self.context = context
+        self.main_model = model = context.model
+
+        self.setSelectionMode(QtWidgets.QAbstractItemView.SingleSelection)
+        self.setHeaderHidden(True)
+        # model
+        self.updated.connect(self.refresh, type=Qt.QueuedConnection)
+        model.add_observer(model.message_updated, self.updated.emit)
+        # UI
+        self._active = False
+        self.list_helper = BuildItem()
+        self.itemDoubleClicked.connect(self.tree_double_clicked)
+
+    def refresh_command(self):
+        self.main_model._update_submodules_list()
+        self.refresh()
+
+    def refresh(self):
+        if not self._active:
+            return
+
+        items = [self.list_helper.get(entry) for entry in
+                 self.main_model.submodules_list]
+        self.clear()
+        self.addTopLevelItems(items)
+
+    def showEvent(self, event):
+        """Defer updating widgets until the widget is visible"""
+        if not self._active:
+            self._active = True
+            self.refresh()
+        return super(SubmodulesTreeWidget, self).showEvent(event)
+
+    def back_to_parent_command(self):
+        # https://stackoverflow.com/questions/7359204/git-command-line-know-if-in-submodule
+        # TODO if stable
+        status, path, _ = self.context.git.git(
+            'rev-parse', '--show-superproject-working-tree')
```

Comment:
You should be able to do `self.context.git.rev_parse(show_superproject_working_tree=True)`.  What version introduced `--show-superproject-working-tree`?  We should probably add a `version.check_git(...)` key for it so that we can only use that flag when Git is new enough.

Reply:
Now the button will be hidden when it is unavailable. 
For submodules in submodules, its .git file contains the top parent instead of its direct parent which makes it a little troublesome.
And I am afraid I need some help about how to monitor the changes of submodules.